src/GitRepository.cc

```diff
@@ -0,0 +1,447 @@
+
+    // commit
+    git_commit_create_v(
+      &commit_id,
+      repo_.get(),
+      "HEAD",
+      my_signature_.get(),
+      my_signature_.get(),
+      "UTF-8",
+      "Initial commit",
+      tree.get(),
+      0
+      );
+
+}
+
+
+void GitRepository::commit(const std::string& commit_message)
+{
+    //get HEAD commit
+    LibGitPointer<git_commit> parent_commit = get_commit();
+    const git_commit* raw_commit = parent_commit.get();
+
+    //define types for commit call and get index
+    LibGitPointer<git_index> index{repository_index(repo_.get())};
+    git_oid tree_id, commit_id;
+    
+    // get tree
+    git_index_write_tree(&tree_id, index.get());
+    LibGitPointer<git_tree> tree{tree_lookup(repo_.get(), tree_id)};
+
+    // create commit
+    int error = git_commit_create(
+        &commit_id,
+        repo_.get(),
+        "HEAD",
+        my_signature_.get(),
+        my_signature_.get(),
+        "UTF-8",
+        commit_message.c_str(),
+        tree.get(),
+        1,
+        &raw_commit
+    );
+
+    if (error) throw task::Error("Cannot commit.");
+}
+
+
+void GitRepository::add()
+{
+    //load index of last commit
+    LibGitPointer<git_index> gindex{repository_index(repo_.get())};
+
+    char *paths[] = {const_cast<char*>("*")};
+    git_strarray array = {paths, 1};
+
+    //add all
+    int error = git_index_add_all(gindex.get(), &array, GIT_INDEX_ADD_DEFAULT, nullptr, nullptr);
+    if (error) throw task::Error("Cannot stage files.");
+
+    // save addition
+    git_index_write(gindex.get());
+}
+
+
+void GitRepository::remove_directory(const std::filesystem::path& seq_directory)
+{
+    //load index of last commit
+    LibGitPointer<git_index> gindex{repository_index(repo_.get())};
+
+    //remove files from directory
+    int error = git_index_remove_directory(gindex.get(), seq_directory.c_str(), 0);
+    if (error) throw task::Error("Cannot remove sequence files.");
+
+    // remove sequence directory from git
+    error = git_index_remove_bypath(gindex.get(), (std::string(seq_directory)+"/").c_str());
+    if (error) throw task::Error("Cannot remove sequence directory.");
+
+    git_index_write(gindex.get());
+
+    // delete sequence
+    std::filesystem::remove_all(repo_path_ / seq_directory);
+}
+
+
+LibGitPointer<git_commit> GitRepository::get_commit(int count)
+{
+    std::string ref = "HEAD^" + std::to_string(count);
+    return get_commit(ref);
+}
+LibGitPointer<git_commit> GitRepository::get_commit(const std::string& ref)
+{
+    git_commit * commit;
+    git_oid oid_parent_commit;
+
+    // resolve HEAD into a SHA1
+    int error = git_reference_name_to_id( &oid_parent_commit, repo_.get(), ref.c_str());
+    if (error) throw task::Error("Cannot find HEAD of branch.");
+
+    // find commit object by commit ID
+    error = git_commit_lookup( &commit, repo_.get(), &oid_parent_commit );
+    if (error) throw task::Error("Cannot find HEAD of branch.");
```

Comment:
With `task::Error` we still intertwine the `git` and `task` namespaces... I think we have two possibilities:
1. Stick everything back into `task::` (at least for the moment, we could do it "right" later)
2. Create a new `git::Error` class and handle the exceptions accordingly
@wyrnat, @MissPeppermintPatty: Opinions?

Reply:
I vote for new git:Error class

Reply:
Okay, I agree. You can probably just copy task::Error.